source/common/formatter/substitution_formatter.h

```diff
@@ -253,8 +255,192 @@ template <class FormatterContext> class FormatterBaseImpl : public FormatterBase
   std::vector<FormatterProviderBasePtr<FormatterContext>> providers_;
 };
 
+// Helper class to parse the Json format configuration.
+class JsonFormatBuilder {
+public:
+  static constexpr absl::string_view BoolFalse = "false";
+  static constexpr absl::string_view BoolTrue = "true";
+  static constexpr absl::string_view NullValue = "null";
+
+  struct JsonString {
+    std::string value_;
+  };
+  struct TmplString {
+    std::string value_;
+  };
+  using FormatterEelements = std::vector<absl::variant<JsonString, TmplString>>;
+
+  /**
+   * Constructor of JsonFormatBuilder.
+   * @param keep_value_type whether to keep the value type in JSON. If false
+   * then all values will be converted to string.
+   */
+  explicit JsonFormatBuilder(bool keep_value_type) : keep_value_type_(keep_value_type) {}
+
+  /**
+   * Convert a proto struct format configuration to a list of raw JSON string and
+   * substitution template string.
+   * Given a proto struct format configuration, it will be parsed to a list of raw
+   * string and substitution template string.
+   *
+   * For example given the following proto struct format configuration:
+   *
+   *   json_format:
+   *     text: "text"
+   *     tmpl: "%START_TIME%"
+   *     number: 2
+   *     bool: true
+   *     list:
+   *       - "list_raw_value"
+   *       - false
+   *       - "%EMIT_TIME%"
+   *     nested:
+   *       text: "nested_text"
+   *
+   * It will be parsed to the following peaces:
+   *
+   *   - '{"text":"text","tmpl":'
+   *   - '%START_TIME%'
+   *   - ',"number":2,"bool":true,"list":["list_raw_value",false,'
+   *   - '%EMIT_TIME%'
+   *   - '],"nested":{"text":"nested_text"}}'
+   *
+   * Finally the substitution template string will be parsed to a list of substitution
+   * formatters. Then join the raw string and output of substitution formatters in order
+   * to get the final output.
+   *
+   * @param struct_format the proto struct format configuration.
+   */
+  FormatterEelements fromStruct(const ProtobufWkt::Struct& struct_format) const;
+
+private:
+  void formatValueToFormatterEelements(const ProtobufWkt::Struct& value) const;
+  void formatValueToFormatterEelements(const ProtobufWkt::Value& value) const;
+
+  const bool keep_value_type_{};
+
+  mutable std::string sanitize_buffer_; // Buffer for sanitizing strings.
+  mutable std::string raw_json_string_; // Buffer for joining raw strings.
+  mutable FormatterEelements output_;   // Final output.
+};
+
+template <class FormatterContext>
+class JsonFormatterImplBase : public FormatterBase<FormatterContext> {
```

Comment:
how does this inheritance work?  I was expecting to see some overrides.

My thesis is that you should be using the JsonStreamer but I might not be understanding how data gets into the access logger.

Reply:
> how does this inheritance work? I was expecting to see some overrides.

I lost `override` of `formatWithContext`. The clang.tidy doesn't complain, weird. Will fix it.

> My thesis is that you should be using the JsonStreamer but I might not be understanding how data gets into the access logger.

See the https://github.com/envoyproxy/envoy/pull/35545#discussion_r1702048822 for my detailed thinking. Basically, if 1) is used, then the JSON streamer will be used to construct JSON output. 

But I think 2) and 3) will provide better performance. We needn't to sanitize the keys, raw values and append delimiters at runtime. The iteration would be simpler and memory friendly (it's iteration of an array). And the adjacent keys, values, and delimiters will be joined to single one raw string at configuration loading time, this would reduce iteration numbers at runtime even further.

Reply:
I don't know if the text formatter is optimal :)

But I see your point about sanitization; the streamer might be less trusting of its input than you are. How do you know you don't need to sanitize the input?

Also I want to lean in on not duplicating complex logic for generating legal json. It'd be better to have that all in json_streamer.cc than duplicate it here. I will note also that you are calling the sanitizer directly yourself; are you sure json_streamer would call it more?

Reply:
> I don't know if the text formatter is optimal :)

Haha, I believe it's yes. Although I think we could do more optimization to the `FormatterProvder`. But it's no business with TEXT or JSON. If we do related optimizations, both TEXT and JSON could benefit from it. This would be another topic/PR.

> But I see your point about sanitization; the streamer might be less trusting of its input than you are. How do you know you don't need to sanitize the input?

Actually, all values, keys will be sanitized.

But the keys, raw values, will be sanitized when loading configuration directly. These sanitized values will be used to construct JSON pieces and be pushed into the `FormatterElements`(an array of JSON pieces and formatter)

Then, at runtime (core data paths), we only need to sanitize output of substitution formatters.

> Also I want to lean in on not duplicating complex logic for generating legal json. It'd be better to have that all in json_streamer.cc than duplicate it here. I will note also that you are calling the sanitizer directly yourself; are you sure json_streamer would call it more?

The core of whole design is to do as much as possible work when loading configuration. The whole JSON constructing is split to two parts. 

The first part, when loading configuration, the keys, raw values, delimiters will be serialized as JSON string pieces (raw JSON strings) directly.
The second part, when flushing log entry, the output of substitution formatter (values that generated by substitution commands) will be sanitized at runtime.
An array (`FormatterElements`)  is used to ensure these JSON string pieces and output of formatter will in the correct position. We only need to join them in order to construct the final complete JSON output.

Reply:
This makes sense. I think it would be best to add methods to JsonSanitizer to stream in pre-sanitized strings, so we don't need to duplicate the logic here.

Reply:
> This makes sense. I think it would be best to add methods to JsonSanitizer to stream in pre-sanitized strings, so we don't need to duplicate the logic here.

SGTM. I can create a help class `JsonSanitizer` to do this for sanitizing short JSON piece.

Reply:
I apologize, I made a critical typo. I didn't want you to change the sanitizer to allow an already sanitized input (I don't see how that would make sense).

I meant to change `JsonStreamer` to accept input that will not be sanitized; just a  new method on the existing class(es).

Reply:
I will say it's a beautiful misunderstanding. I have implement a helper `JsonSanitizer` to construct short JSON piece. It works will and make the code more clean with only minor performance penalty compare to initial version of the PR. (I didn't find why it bring the penalty, I can only blame it on compiler optimization now)

```
-------------------------------------------------------------------------------
Benchmark                                     Time             CPU   Iterations
-------------------------------------------------------------------------------
BM_AccessLogFormatterSetup                 9013 ns         9011 ns        79050
BM_AccessLogFormatter                       494 ns          494 ns      1412629
BM_StructAccessLogFormatter                7718 ns         7715 ns        87465
BM_TypedStructAccessLogFormatter           6364 ns         6362 ns       112444
BM_LegacyJsonAccessLogFormatter           13404 ns        13399 ns        53254
BM_LegacyTypedJsonAccessLogFormatter      10904 ns        10900 ns        60094
BM_JsonAccessLogFormatterNew                568 ns          567 ns      1237317
BM_TypedJsonAccessLogFormatterNew           767 ns          767 ns       918201
BM_FormatterCommandParsing                 40.6 ns         40.5 ns     17215937

```

**Although I think `JsonStreamer` is great. It doesn't always fit all scenarios.**
First, it depends on Buffer::Instance. This is unnecessary for log lines where are not a very large string. It only bring addtionnal overhead. Here is a quick result by replace the short string buffer to `Buffer::OwnedImpl` (~35% degration):

```
-------------------------------------------------------------------------------
Benchmark                                     Time             CPU   Iterations
-------------------------------------------------------------------------------
BM_AccessLogFormatterSetup                 8993 ns         8979 ns        78521
BM_AccessLogFormatter                       494 ns          493 ns      1425589
BM_StructAccessLogFormatter                7823 ns         7811 ns        91837
BM_TypedStructAccessLogFormatter           6288 ns         6279 ns       110759
BM_LegacyJsonAccessLogFormatter           13576 ns        13428 ns        52173
BM_LegacyTypedJsonAccessLogFormatter      12021 ns        12008 ns        60285
BM_JsonAccessLogFormatterNew                766 ns          766 ns       923385
BM_TypedJsonAccessLogFormatterNew           912 ns          911 ns       765203
BM_FormatterCommandParsing                 41.3 ns         41.2 ns     16977352
```

Second, there are some un-common requirements when format the log line. That will bring big effect to the `JsonStreamer`'s APIs.
For example, we may hope to handle the quoting manually when handle the multiple commands substitution formatter. (By this way, we can avoid joining the output of multiple command providers to single string. We can push output to the buffer directly.)

Reply:
So, I will vote to the new `JsonSanitizer` as a light and quick helper to construct short JSON piece.